MPL sockaddr interface.

hzhou opened the pull request:
This is a result of reviewing this issue:
    https://github.com/pmodels/mpich/issues/2889

Socket handling codes are mostly trivial but messy, and there are redundant code duplicated in multiple place. This is an attempt to centralize the code. It also tries to bring in IPv6 compatibility.

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
test:jenkins/ch3/tcp

raffenet commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,283 @@
+    struct addrinfo ai_hint;
+    struct addrinfo *ai_list;
+    int ret;
+    memset(p_addr, 0, sizeof(*p_addr));
```
Style wise I would prefer to see a blank line between these declarations and the first statement.

raffenet commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,283 @@
+        return 1;
+    }
+    if (af_type == AF_INET) {
+        struct sockaddr_in *p_addr4 = (struct sockaddr_in *) p_addr;
```
Same preference for blank line after declarations inside if/else blocks.

yfguo commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,316 @@
+    } else if (af_type == AF_INET6) {
+        memcpy(p_addr, ai_list->ai_addr, sizeof(struct sockaddr_in6));
+    } else {
+        ret = 1;
```
I prefer to abort here since `af_type` is having an unexpected value here. Something must be really wrong if we hit this branch.

raffenet replied:
Remember this is down in MPL, not MPI land. It should be the responsibility of MPICH to abort where there is a failure.

hzhou replied:
Yanfei is referring to the case that should not happen or is simply error.  I  am replacing these cases with simply assert(0).

yfguo commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,316 @@
+int MPL_get_sockaddr_direct(int type, struct sockaddr_storage *p_addr)
+{
+    memset(p_addr, 0, sizeof(*p_addr));
+    if (type < 0 || type > 1) {
```
Please define the value of `type` in an enum constants, so that we are not putting magic numbers in the code.

yfguo commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,316 @@
+{
+    memset(p_addr, 0, sizeof(*p_addr));
+    if (type < 0 || type > 1) {
+        return 1;
```
Similarly, we should define the return value using enum constants. Or better yet, we can just abort here.

yfguo commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,316 @@
+        }
+        return 0;
+    } else {
+        return 1;
```
Similar to previous places of checking `af_type`, better be aborting here.

yfguo commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,316 @@
+    }
+    freeifaddrs(ifaddr);
+    if (!found) {
+        return -1;
```
The return value needs to be defined as an enum constant.

hzhou replied:
The idea is to serve as wrapper and mimic the behavior of getifaddr, as well as the rest of the posix socket convention -- that return -1 on error and application checks errno for details.

yfguo replied:
OK. I see. Then, I have no problem

yfguo commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,316 @@
+    } else {
+        MPL_get_sockaddr_direct(MPL_SOCKADDR_ANY, &addr);
+    }
+    ret = -1;
```
Init `ret` at definition.

yfguo commented on src/mpl/src/sock/mpl_sockaddr.c:
```diff
@@ -0,0 +1,316 @@
+        } else if (errno == EADDRINUSE) {
+            continue;
+        } else {
+            return -1;
```
Define the return value as an enum constant.

yfguo commented:
test:jenkins/ch3/tcp

yfguo commented:
@raffenet I think this one is good to go. I don't feel strongly about the suggested changes to the return values.

raffenet commented:
Looks like there's a conflict to resolve. If this is approved, then please signoff and merge.

raffenet commented:
Re-test to make sure my conflict fix was correct. Otherwise this is ready.
test:jenkins/ch3/most

raffenet commented:
Looks like `ch3/sock` configurations are broken with this. Needs a fix.

hzhou commented:
test:jenkins/ch3/sock

hzhou commented:
test:jenkins/ch3/sock

hzhou commented:
test:jenkins/ch3/sock

hzhou commented:
test:jenkins/ch3/most

hzhou commented:
test:jenkins/ch3/most

hzhou commented:
test:mpich/warnings

hzhou commented:
test:mpich/warnings

hzhou commented:
test:mpich/warnings

hzhou commented:
test:jenkins/ch3/most

hzhou commented:
@raffenet I have updated this PR. It has been a while. Could you take one more look?

hzhou commented:
> This makes `getaddrinfo` required to build MPICH, correct? Just making we understand the baseline.

I think so. From the manpage, seems it is conforming to `_POSIX_C_SOURCE >= 200112L`.

hzhou commented:
test:jenkins/ch3/most

hzhou commented:
test:mpich/warnings

hzhou commented:
test:jenkins/ch3/tcp

hzhou commented:
> I think this is pretty much ready to go. The abstraction allows us to add a fallback implementation, if needed. However this functionality should be everywhere we intend to run, so I don't expect to run into issues. One last question - have we dug up all the old issues this fixes and added links in commit messages or PR description? E.g. localhost lookup for laptop users, ipv6 support, etc?

Added ipv6 issue and pr links to the description. I'll make another PR to fix mac os `.local` dns issue once this is merged.
